rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,588 @@
+    ),
+}
+```
+
+> Note: In the above configuration and along this RFC configuration variables are prefixed with `RDM_`, this means that custom fields are being added to _records_. However, the same configuration variables are available in communities and will be used then adding custom fields to them. Namely, `COMMUNITIES_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS_UI`.
+
+All the available custom field types will be implemented in `<Type>CF` classes. These will be in charge of defining the properties that reflect the data model, such as Marshmallow schema and Elasticsearch mapping. In the previous example `VocabularyCF`.
+
+```python
+class BaseCF(ABC):
+"""Base Custom Field class."""
+
+    def __init__(self, name):
+        """Constructor."""
+        super().__init__()
+        self.name = name
+
+    @property
+    @abstractmethod
+    def mapping(self):
+        """Return the mapping."""
+        pass
+
+    @property
+    @abstractmethod
+    def schema(self):
+        """Marshmallow schema for custom fields."""
+        pass
+
+    @property
+    def ui_schema(self):
+        """Marshmallow UI schema for custom fields.
+
+        Controls how the field will be serialized in the UI when using the
+        `UIJSONSerializer`. If not overriden, the default implementation will use
+        the `BaseCF.schema` property.
+        """
+        return self.schema
+
+class TextCF(BaseCF):
+"""Text custom field."""
+
+    @property
+    def mapping(self):
+        return {"type": "text"}
+
+    @property
+    def schema(self):
+        return SanitizedUnicode()
+```
+
+Note that all custom fields will be added inside the `custom` parent field, at the same level than `metadata`.
+
+```json
+{
+   "metadata": {
+        ...
+    },
+    "custom_fields": {
+        "experiment": ...
+    }
+}
+```
+
+#### Custom field type / Marshmallow / Elasticsearch chart
+
+The following table represents the type conversion/relation between custom field type, Marshmallow fields and Elasticsearch mapping.
+
+| Custom Field type | Marshmallow field | Elasticsearch mapping |
+| :---------------: | :---------------: | :-------------------: |
+|      string       | SanitizedUnicode  |         text          |
+|      keyword      | SanitizedUnicode  |        keyword        |
+|       date        |     DateField     |    isoformat date     |
+|     date range    |   EDTFDateString  |  multi isoformat date |
+|      integer      |      Integer      |        integer        |
+|      double       |      Number       |        double         |
+|      boolean      |      Boolean      |        boolean        |
+|       list        |       List        |   inner value type    |
+|       dict        |       Dict        |        object         |
+
+Note the date range is handled as other date ranges using the elasticsearch dumper (`EDTFListDumperExt`) to extract the gte (lower) and lte (upper) date limits.
+
+#### Data validation
```

Comment:
Minor: How do I provide a custom validation function? E.g. I want my integer to be a number between 0 and 100.

Reply:
Discussion: OK, add

Reply:
Can you customize/override error messages?